include/services/internal/hash_table.h

```diff
@@ -0,0 +1,133 @@
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef __DAAL_ONEAPI_INTERNAL_HASH_TABLE_H__
+#define __DAAL_ONEAPI_INTERNAL_HASH_TABLE_H__
+
+#include "services/daal_string.h"
+#include "services/internal/error_handling_helpers.h"
+#include "services/error_indexes.h"
+
+#include <iostream>
+
+
+namespace daal
+{
+namespace services
+{
+namespace internal
+{
+
+template<class T, size_t size>
+class HashTable
+{
+public:
+    HashTable()
+    {
+        for (size_t i = 0; i < size; ++i)
+            _arr[i] = nullptr;
+    }
+
+    ~HashTable()
+    {
+        for (size_t i = 0; i < size; ++i)
+        {
+            delete _arr[i];
+            _arr[i] = nullptr;
+        }
+    }
+
+    bool contain(services::String key, services::Status &status)
+    {
+        size_t id = get_unique_hash(key, status);
+        if (!status.ok())
+            return false;
+        return _arr[id] != nullptr;
+    }
+
+    void add(services::String key, services::SharedPtr<T> obj, services::Status &status)
+    {
+        size_t id = get_unique_hash(key, status);
+        if (!status.ok())
+            return;
+        if (_arr[id] == nullptr)
+        {
+            _arr[id] = new Entry();
+            _arr[id]->key = key;
+            _arr[id]->obj = obj;
+            return;
+        }
+        status.add(services::ErrorExecutionContextCache);
+    }
+
+    services::SharedPtr<T> get(services::String key, services::Status &status)
+    {
+        size_t id = get_unique_hash(key, status);
+        if (!status.ok())
+            return services::SharedPtr<T>();
+
+        if (_arr[id] != nullptr)
+            return _arr[id]->obj;
+
+        status.add(services::ErrorExecutionContextCache);
+        return services::SharedPtr<T>();
+    }
+
+private:
+    struct Entry
+    {
+        services::SharedPtr<T> obj;
+        services::String key;
+    };
+    Entry* _arr[size];
+
+    const static size_t magic_p = 5381;
+    static size_t get_hash(services::String key)
+    {
+        size_t hash = 0;
+        size_t magic_p_pow = 1;
+        for (size_t i = 0; i < key.length(); ++i)
+        {
+            hash += (key[i] + 1) * magic_p_pow; // it's OK
+            magic_p_pow *= magic_p; // it's OK
+        }
+        return hash;
+    }
+
+    size_t get_unique_hash(services::String key, services::Status &status)
+    {
+        size_t id = get_hash(key) % size;
+        const size_t start_id = id;
+        while (_arr[id] != nullptr && (_arr[id]->key.length() != key.length() ||
+            _arr[id]->key.length() == key.length() && strncmp(_arr[id]->key.c_str(), key.c_str(), key.length()) != 0))
```

Comment:
Condition is very bulky, suggest to implement appropriate `String` comparison operator, if there is no any.